### DISCUSSION TITLE
add support for Bosch/ HPC410 "primary heating pump mode"

### DISCUSSION THREAD

**Discussion by derSchreifritz** [emsesp/EMS-ESP32#1449] [Ideas]
hi, can you add support for the "primary heating pump mode"?
Options are Automatic or ON.
Description reads: Select automatic or continuous mode for primary heating pump.
I attach system info and a log.
I changed the setting from On to Automatic and back 3 times.
Hope this makes sense.
[emsesp_info.json](https://github.com/emsesp/EMS-ESP32/files/13427949/emsesp_info.json)
[log(10).txt](https://github.com/emsesp/EMS-ESP32/files/13427950/log.10.txt)

**Comment by MichaelDvP** [comment]
> I changed the setting from On to Automatic and back 3 times.

6 changes in 25 seconds? There is nothing change in the log. You need to give the bus some time to set the change, always wait a minute for a change become active before you change again.

**Reply by derSchreifritz** [reply]
hope this is better:
[log(12).txt](https://github.com/emsesp/EMS-ESP32/files/13431201/log.12.txt)
changed from Auto to On (continuous), back to auto, and lastly to On.

**Reply by MichaelDvP** [reply]
> hope this is better:

Yes: 0x48B, offset 18,, added to my test build

**Reply by derSchreifritz** [reply]
yes, thanks, I found it and it's working: select.boiler_hppumpmode Primary heatpump mode
Maybe name Primary Circulation pump mode would be better to avoid confusion.

**Reply by MichaelDvP** [reply]
`Circulation pump` is normally used for the DHW circulation, here we have heating (i think so, don't know where this pump is located in hydraulic scheme). What name is used on the panel, please check for different languages.

**Reply by derSchreifritz** [reply]
this is indeed about the internal circulation pump, inside the unit. This one is also reflected by Circulation pump speed when active (heating, cooling or dhw run). During inactivity, Circulation pump speed sensor always goes to 10%, but depending on our Primary heatpump mode there is I think the same pump running slowly or not at all. Continous circulation is recommended when there is no external pump, but in my case not needed in summer.
 I don't have another (external) circulation pump, so all the dhw circulation entities do nothing here.
In English on the panel, Primary heatpump mode is found under service menu: Heatpump: Pumps: Prim. heating pump mode.
"heating pump" I find a bit confusing here. In Dutch the panel says "cv-pomp", German "Heizungspumpe", French "pompe chauffage primaire"

**Comment by nicoleise** [comment]
I (too) find the naming very confusing, and would suggest that the ID is added to pumps and valves, similar to temperature sensors, and that the parameters for the PC0 pump be renamed. 

**If** these parameters define the behavior of the internal circulation pump for the heating side of the heatpump/boiler (PC0), then it really should be labelled circulation pump - just like the feedback object for the same pump, which is "circulation pump speed". 

I have actually been looking for this exact option for some time, but thought it didn't exist simply because of the name. I've skipped over it at least 20 times, thinking it somehow overrides the entire heatpumps modulation and behavior. 

The brine pump in heatpumps should be PB1 for horisontal collectors and PB3 for vertical collectors (drilled). 
The circulation pumps for the heating system should be PC0 for the internal circulation pump and PC1 for the optional, exterior circulation pump (for example after a bypass if a bypass is installed). PC1 is present and more properly named in the parameters already. 


All that being said, when I test these options, they seem to do nothing. Are they perhaps of lower priority than the heatpumps own demands? I ask because if I change them, nothing happens, except for the mode, which reverts back to auto after being set to continuous. But my heatpump is also currently running in heating mode. Can anyone confirm whether these parameters are only in effect when the boiler/heatpump isn't "on" and has demand for the pump?

I'm wondering if these parameters perhaps correspond to the parameters under the installer menu/Diagnostics/Function Tests, in which case they are likely only active after first enabling "Manual control" to allow access to the function tests in the HMI?

**Reply by MichaelDvP** [reply]
Yes, the naming is always difficult. We developers do not have the equipment, new entities are requested by users that help to log and identify the value in the bus telegrams. These users tell us the names, based on the display or manual thes have. Sadly Bosch changes names often and same parameter in different controllers (RC300, RC310, BC400, UI800) can have different names. Also names changes dependent on software version or brand.  So emsesp shows only the name the user who developt the value told us. 
> The brine pump in heatpumps should be PB1

We can add the short name if we know the entity. The unique name is in brackets in customization and in mqtt publish. I don't know what entity `the brine pump` is, so without unique name i can not rename.

Best way: Take a look at https://github.com/emsesp/EMS-ESP32/blob/dev/src/core/locale_translations.h
identify the entities you like to add the clamp name like `BR1`, edit it and make a PR.

**Reply by nicoleise** [reply]
> I don't know what entity `the brine pump` is, so without unique name i can not rename.

The entity is hpbrinepumpspd, and it's line 418 in the linked file. I cannot say for sure if Bosch uses the same address/register for both the PB1 and PB3 pumps though. In other words, this could in theory be the regulator output which then gets fed to either or both. 

I don't mind contributing to the file, but two questions:
1 - is there some naming convention?
2 - I assume changing the mqtt topic results in breaking changes, but wouldn't that also be the case for translations? For example if users look for a certain payload and we change the translation of it?


Finally, I tested the parameters and as I suspected they correspond to the function test menu options. I also verified that if I enable "manual control" in the function test menu to "unlock" the ability to control the heatpump, the entities work as expected. If the option "manual control" is not set, the entities do not work. Leaving the heatpump in manual control is not an option as it's own controls are disabled. So for modern devices using the HMI800, they are of no real use. 

The PC0 pump that I am looking to control is managed by the heatpump using an internal Modbus connection. 

I'd like to be able to set it to run continuously, ideally also to specify the modulation/speed, when the heatpump is not using it. Is there something I can do to enable the end gateway to do this? Its not a parameter defined by Bosch (outside of the function test menu), but obviously a register containing these things must exist.

**Reply by MichaelDvP** [reply]
> I don't mind contributing to the file, but two questions:
1 - is there some naming convention?
2 - I assume changing the mqtt topic results in breaking changes, but wouldn't that also be the case for translations? For example if users look for a certain payload and we change the translation of it?

1 - mqtt names has to be unique and lowercase, shorter names are better.
2 - yes, changing mqtt names is a breaking change. Discovery service also uses the mqtt names for unique id, so renaming long names and the translations is not breaking.
